Added MessageHub polling to refresh Elasticsearch cache

fabiogomezdiaz opened the pull request:
- Added MessageHub topic subscription and message polling capability  …
- If reload_cache message found, then loads updated database into Elasticsearch
- Added configuration file to put MessageHub and Elasticsearch credentials and run locally
- Added scripts to load MessageHub and Elasticsearch credentials into environment variables
- Added checks for Message Hub and Elasticsearch credentials before app starts polling
- Updated gradle docker task and startup.sh script to load configuration into docker
- Updated dockerfile to copy config.json and load_config.sh files
- Updated docks to run application locally, in docker, and in Bluemix
- TODO: Flow diagram and Elasticsearch container deployment instructions

ssibm committed:
devops pipeline, unit test, contract validation (#13) (#14)

* Merge from upstream (#1)

* initial support for New Relic

* Add circuit breaker to getAll items

* get app name from CG_NAME

* use ROUTE_HOSTNAME to notate devices in new relic

* use CG_NAME to notate devices in new relic

* Fix the hystrix command type casting issue

* instruction to unmap public route

* move mysql into local repo

* add unit tests

* fix REST API

* add bluemix devops pipeline temporarily

* rename pipeline.yaml to pipeline.yml

* fix error on POST

* contract validation

* updated pipeline

* fix vulnerability advisor

* toolchain

* fix deploy.json

* fix variable substitutions

* fix image name variables

* fix more variables

* fix mysql deployment

* change route for contract validation instances

* deploy test container to a different route and container group name

* fix context path

* add vpn and mysql pipelines

* fix mysql pipeline

* fix mysql pipeline?

* fix toolchain?

* mysql pipeline

* mysql pipeline

* fix toolchain?

* mysql pipeline

* mysql pipeline

* add back mysql connection info

* fix deployment to mysql test container

* asdgsda

* variablize jdbc url

* auto deploy hybrid app

* fix context path

Commit:
- Added MessageHub topic subscription and message polling capability
- If reload_cache message found, then loads updated database into Elasticsearch
- Added configuration file to put MessageHub and Elasticsearch credentials and run locally
- Added scripts to load MessageHub and Elasticsearch credentials into environment variables
- Added checks for Message Hub and Elasticsearch credentials before app starts polling
- Updated gradle docker task and startup.sh script to load configuration into docker
- Updated dockerfile to copy config.json and load_config.sh files
- Updated docks to run application locally, in docker, and in Bluemix
- TODO: Flow diagram and Elasticsearch container deployment instructions

Commit:
- Updated docks to suggest provisioning Searchly while Elasticsearch container instructions are ready
- Typo in the build.gradle

Commit:
Merge branch 'integration' of https://github.com/ibm-cloud-architecture/refarch-cloudnative-micro-inventory

fabiogomezdiaz commented on config.json:
```diff
@@ -0,0 +1,22 @@
+  "mh_message": "refresh_cache",
+  "VCAP_SERVICES": {
+    "messagehub": [{
+      "credentials": PASTE_MESSAGE_HUB_CREDENTIALS_OBJECT_HERE,
```
This is intentional so you just copy and paste from Message Hub Bluemix credentials

fabiogomezdiaz commented on docker/Dockerfile:
```diff
@@ -14,6 +14,10 @@ RUN apt-get update && apt-get install -y jq
 # Copy agent files
 COPY agents/ /agents/
 
+# Copy configuration files
```
Added this to make it easy to run it in docker

fabiogomezdiaz commented on docker/startup.sh:
```diff
@@ -6,6 +6,7 @@ export JAVA_OPTS="-Djava.security.egd=file:/dev/./urandom"
 
 # Load agent support if required
 source ./agents/newrelic.sh
+source load_config.sh
```
Complements the above

fabiogomezdiaz commented on load_config.sh:
```diff
@@ -0,0 +1,18 @@
+#!/usr/bin/env bash
```
To make it easy to load MessageHub and Elasticsearch credentials and run locally and in docker

fabiogomezdiaz commented on src/main/java/inventory/mysql/Application.java:
```diff
@@ -15,5 +15,26 @@
         ApplicationContext ctx = SpringApplication.run(Application.class, args);
         System.out.println("Inventory microservice is ready for business...");
+
+        String connection = System.getenv("es_connection_string");
```
Checks to make sure we have what we need before starting

fabiogomezdiaz commented on src/main/java/inventory/mysql/ElasticSearch.java:
```diff
@@ -0,0 +1,94 @@
+            // TODO: Use some SpringBoot data approach to get the data
+            OkHttpClient client = new OkHttpClient();
+            Request request = new Request.Builder()
+            .url("http://localhost:8080/micro/inventory")
```
I know! Best I could come up with in the meantime

fabiogomezdiaz commented on src/main/java/inventory/mysql/ElasticSearch.java:
```diff
@@ -0,0 +1,94 @@
+package inventory.mysql;
```
This file will get all the rows from database and update the elasticsearch cache

fabiogomezdiaz commented on src/main/java/inventory/mysql/MHConsumer.java:
```diff
@@ -0,0 +1,176 @@
+package inventory.mysql;
```
This file setups the JAAS configuration before talking with MessageHub. Then it starts polling topic

fabiogomezdiaz commented on src/main/resources/application.yml:
```diff
@@ -3,7 +3,7 @@ eureka:
      preferIpAddress: true
```
Oops, need to take this out

Commit:
Don't tell anyone

Commit:
No longer need this

Commit:
- Changed MYSQL data model
- Updated Inventory class
- Updated Inventory tests

Commit:
- Made it so InventoryController can be used to query data when updating Elasticsearch
- Updating Message Hub polling
- Fixed some typos on defaults for ElasticSearch and Message Hub

Commit:
- Fixed some issues with running out of memory heap

Commit:
- Added instructions to deploy Elasticsearch container both locally and in Bluemix

gangchen03 commented:
01/31 Review:
1. API definition of Inventory needs refinement, for example shouldn't have /update, /delete in REST path.
2. pre-requisite, for Docker, we shouldn't use the Docker-machine any more, use the Docker 1.12 or later. Use this link instead https://docs.docker.com/engine/installation/
3. ElasticSearch step 3, Copy the following as your ElasticSearch connection string what does it mean? Please explain, for example, the endpoint to communicate with ElasticSearch REST API?
4. MySQL load-data.sh script failed the first time, maybe add instruction to wait for MySQL fully started, 
5. Why is there a mysql folder under the micro-inventory project?
6. Run Inventory Service application on Localhost, please indicate which directory reader will be working under and,  give a complete example of the config.json
7. Validate the Running app locally, clearly mentioned which is the command, which is the result. They are now seems to be a single command.
8. Getting MessageHub Error
9. Config.json should be merged with application.yml file

jkwong888 commented:
My comments:
- I prefer to put the pre-requisites up-front, i.e. provision elasticsearch, mysql, and messagehub before starting on the build/deployment steps
- I prefer to not use that mysql repo.  The repo doesn't really add much value, as it's just the mysql image from docker hub with an extra script for the schema/data load.  Can we just document this instead:
  -  a docker pull mysql:latest from dockerhub
  - a docker tag and docker push into bluemix container image registry
  - cf ic run the mysql container
  - cf ic cp the script into the container
  - cf ic exec the script in the container

we can use compose for mysql (when it comes out of beta), or an on-prem mysql instance in production

Can we use the schema from the mysql repo and then deprecate or retire the mysql repo?

Commit:
- Created Config class to handle config propagation, it pull from the application.yml
- Got rid of code that handled the old config.json
- Added code to create message hub topic only if it does not exist

Commit:
New application.yml properties

Commit:
Gradle stuff

Commit:
- Updated docs
- Removed some unnecessary config variables

Commit:
- Now it detects whether message hub is bound to bluemix and gets credentials that way if so
- Update docs with configuration steps details

Commit:
- Improved docs for container commands
- Check different paths for config file
- Improved mysql load data script

Commit:
- Forgot to remove
